elasticdl/python/master/pod_manager.py

```diff
@@ -418,153 +473,154 @@ def _event_cb(self, event):
-            if evt_type == "MODIFIED" and phase == "Succeeded":
-                for callback in self._pod_event_callbacks:
-                    callback.on_pod_succeeded(
-                        PodInfo(type=None, id=None, name=pod_name),
-                        ClusterContext(pod_manager=self),
-                    )
-
-        if relaunch_worker and worker_id >= 0:
-            logger.info("Relaunching worker.")
             new_worker_id = self._next_worker_id_fn()
             with self._lock:
                 self._worker_pod_priority[
                     new_worker_id
-                ] = self._worker_pod_priority[worker_id]
+                ] = self._worker_pod_priority[pod_id]
             self._start_worker(new_worker_id)
-        elif relaunch_ps:
-            logger.info("Relaunching ps.")
-            # Note: the ID and service address for relaunched parameter
-            # server are intentionally left unchanged to support fault
-            # tolerance.
-            self._start_ps(ps_id)
-
-    def check_all_workers_exited(self):
-        workers_exited = []
-        for (_, _, phase,) in self._worker_pods_ip_phase.values():
-            workers_exited.append(
-                phase == PodStatus.FAILED or phase == PodStatus.SUCCEEDED
+
+    @staticmethod
+    def get_pod_state_flow(from_status, event_type, phase):
+        for pod_state_flow in POD_STATE_FLOWS:
+            if (
+                from_status == pod_state_flow.from_status
+                and event_type == pod_state_flow.event_type
+                and (
+                    phase == pod_state_flow.phase
+                    or pod_state_flow.phase is None
+                )
+            ):
+                return pod_state_flow
+
+        return None
+
+    @property
+    def all_workers_exited(self):
+        with self._lock:
+            all_exited = all(
+                [
+                    pod_info.status
+                    in [
+                        PodStatus.SUCCEEDED,
+                        PodStatus.FAILED,
+                        PodStatus.DELETED,
+                    ]
+                    for pod_info in self._pod_info_cache[
+                        PodType.WORKER
+                    ].values()
+                ]
             )
-        self.all_workers_exited = all(workers_exited)
+
+        return all_exited
 
     def get_alive_workers(self):
-        alive_workers = []
-        for pod_name, _, phase in self._worker_pods_ip_phase.values():
-            if phase == PodStatus.RUNNING:
-                alive_workers.append(pod_name)
-        return alive_workers
+        with self._lock:
+            return [
+                pod_info
+                for pod_info in self._pod_info_cache[PodType.WORKER].values()
+                if pod_info.status == PodStatus.RUNNING
+            ]
+
+    def get_alive_worker_name_addr(self):
+        alive_workers = self.get_alive_workers()
+        alive_workers.sort(key=lambda tup: tup.id)
```

Comment:
It is not suitable to sort workers by id. The master will submit workers at the start and the worker will start at different times.  We cannot guarantee the worker with id=0 starts to run before the worker with id=1.

Reply:
We can sort workers by start time in the next PR.